### ISSUE TITLE
Consider combining some of the crates in this repo & changing how intra and inter crate dependencies are used

### ISSUE THREAD

**Issue by emilyaherbert** [FuelLabs/sway#1887]
When a crate exports a type as `pub`, clippy is not able to detect whether or not it is unused, from my understanding. The `sway-core` crate exports _a lot_ of things as `pub`, and I suspect that some portion of those exports do not need to be `pub` and instead could be `pub(crate)`, `pub(super)`, or private. In a code base as large as this one, this can create some problems---code can be dead and go undetected for some time:
https://github.com/FuelLabs/sway/pull/1886
https://github.com/FuelLabs/sway/pull/1831#pullrequestreview-994365737

I don't know of the particular reason for the parser, IR, etc, to be split into different crates---are we expecting users to use `sway-parse`, `sway-ir`, etc, directly? If not, we should consider combining these into the `sway` crate so that we can update the dependencies as described above.

**Comment by otrho** [comment]
Providing a clearer interface between the Sway modules is definitely worth fixing.  And whether then moving the IR and parser into modules rather than crates is debatable.

_Ideally_ the parser remains separated so that the tools like LSP or the formatter can use it without needing all of Sway.  Similarly, `sway-ir` should really be called `fuel-ir` and actually not be tied to Sway directly.  _Ideally_ we'd be able to use `fuel-ir` as a middle end for other languages.

**Comment by mitchmindtree** [comment]
I agree with @otrho, but I also see where you're coming from @emilyaherbert.

*Remembers my first `sway-core` PR and removing the `dependency_graph` from all of type-checking that wasn't used anywhere but wasn't detected as `TypeCheckArguments` was `pub` for some reason.*

I think, like you said, this more-so hints at the need to be much more conservative in what we expose as `pub` from `sway-core` more generally. We should default to private if we can, `pub(crate)`/`pub(super)` if we need it, and only use `pub` for items that we are comfortable in supporting public usage of.

Another motivation to be more conservative is that the more things are `pub`, the more potential surface for breakage when considering new SemVer releases.
